### PR TITLE
Fix tablefilter.RandomSubset, add optional pass-through for len < n

### PR DESCRIPTION
**Pull request by barentine**
Addresses issue #appears this module has never actually worked.
image
Screen Shot 2020-12-30 at 2 56 00 PM

**NOTE**
not trying to be snarky by including the blame, but want to point out we might get away with changing the traits names this time

**Is this a bugfix or an enhancement?**
both now
**Proposed changes:**
- don't attribute error instead of putting the output into the namespace
- while we have the flexibility to rename things, do a little of that
- add flag to allow pass through in cases when the input length is shorter than the selected number
- add unit test

### CODE REVIEW

**Comment by barentine** [comment]
use case for pass-through is in detection recipe module, if I only want to image a max of 100 cells or so, but if I only detect 99 I still want to image them

File: PYME/recipes/tablefilters.py [hunk 1]
```diff
@@ -179,21 +179,24 @@ def execute(self, namespace):
 @register_module('RandomSubset')
 class RandomSubset(ModuleBase):
     """Select a random subset of rows from a table"""
-    input = Input('input')
-    output = Output('output')
-    numToSelect = Int(100)
+    input_name = Input('input')
```
Comment (David-Baddeley) [inline PYME/recipes/tablefilters.py]:
```suggestion
    input = Input('input')
```

File: PYME/recipes/tablefilters.py [hunk 2]
```diff
@@ -179,21 +179,24 @@ def execute(self, namespace):
...
+    output_name = Output('output')
```
Comment (David-Baddeley) [inline PYME/recipes/tablefilters.py]:
```suggestion
    output = Output('output')
```

File: PYME/recipes/tablefilters.py [hunk 3]
```diff
@@ -179,21 +179,24 @@ def execute(self, namespace):
...
+    n_to_select = Int(100)
+    require_at_least_n = Bool(False)
     
     def execute(self, namespace):
-        data = namespace[self.input]
+        data = namespace[self.input_name]
```
Comment (David-Baddeley) [inline PYME/recipes/tablefilters.py]:
```suggestion
        data = namespace[self.input]
```

File: PYME/recipes/tablefilters.py [hunk 4]
```diff
@@ -179,21 +179,24 @@ def execute(self, namespace):
...
-            out.mdh = data.mdh
+            out.mdh = MetaDataHandler.DictMDHandler(data.mdh)
         except AttributeError:
             pass
 
-        namespace[self.outputName] = out
-        
-
-
+        namespace[self.output_name] = out
```
Comment (David-Baddeley) [inline PYME/recipes/tablefilters.py]:
```suggestion
        namespace[self.output] = out
```

File: tests/PYME/recipes/test_tablefilters.py [hunk 5]
```diff
@@ -0,0 +1,15 @@
+
+from PYME.recipes import tablefilters
+
+def test_random_selection():
+    from PYME.IO import tabular
+    import numpy as np
+    d = tabular.DictSource({'test': np.arange(100)})
+
+    out = tablefilters.RandomSubset(n_to_select=5,
+                                    require_at_least_n=True).apply_simple(input_name=d)
```
Comment (David-Baddeley) [inline tests/PYME/recipes/test_tablefilters.py]:
```suggestion
                                    require_at_least_n=True).apply_simple(input=d)
```

File: tests/PYME/recipes/test_tablefilters.py [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+    assert len(out) == 5
+    
+    out = tablefilters.RandomSubset(n_to_select=150,
+                                    require_at_least_n=False).apply_simple(input_name=d)
```
Comment (David-Baddeley) [inline tests/PYME/recipes/test_tablefilters.py]:
```suggestion
                                    require_at_least_n=False).apply_simple(input=d)
```

File: PYME/recipes/tablefilters.py [hunk 7]
```diff
@@ -181,19 +181,22 @@ class RandomSubset(ModuleBase):
     """Select a random subset of rows from a table"""
     input = Input('input')
     output = Output('output')
-    numToSelect = Int(100)
+    n_to_select = Int(100)
```
Comment (David-Baddeley) [inline PYME/recipes/tablefilters.py]:
```suggestion
    num_to_select = Int(100)
```

File: PYME/recipes/tablefilters.py [hunk 8]
```diff
@@ -181,19 +181,22 @@ class RandomSubset(ModuleBase):
...
+    require_at_least_n = Bool(False)
     
     def execute(self, namespace):
         data = namespace[self.input]
         
-        out = tabular.RandomSelectionFilter(data, num_Samples=self.numToSelect)
+        if self.require_at_least_n:
+            n_to_select = self.n_to_select
+        else:
+            n_to_select = min(len(data), self.n_to_select)
```
Comment (David-Baddeley) [inline PYME/recipes/tablefilters.py]:
```suggestion
            n_to_select = min(len(data), self.num_to_select)
```

File: PYME/recipes/tablefilters.py [hunk 9]
```diff
@@ -181,19 +181,22 @@ class RandomSubset(ModuleBase):
...
-    numToSelect = Int(100)
+    n_to_select = Int(100)
+    require_at_least_n = Bool(False)
     
     def execute(self, namespace):
         data = namespace[self.input]
         
-        out = tabular.RandomSelectionFilter(data, num_Samples=self.numToSelect)
+        if self.require_at_least_n:
+            n_to_select = self.n_to_select
```
Comment (David-Baddeley) [inline PYME/recipes/tablefilters.py]:
```suggestion
            n_to_select = self.num_to_select
```

File: tests/PYME/recipes/test_tablefilters.py [hunk 10]
```diff
@@ -0,0 +1,15 @@
+
+from PYME.recipes import tablefilters
+
+def test_random_selection():
+    from PYME.IO import tabular
+    import numpy as np
+    d = tabular.DictSource({'test': np.arange(100)})
+
+    out = tablefilters.RandomSubset(n_to_select=5,
```
Comment (David-Baddeley) [inline tests/PYME/recipes/test_tablefilters.py]:
```suggestion
    out = tablefilters.RandomSubset(num_to_select=5,
```

File: tests/PYME/recipes/test_tablefilters.py [hunk 11]
```diff
@@ -0,0 +1,15 @@
...
+                                    require_at_least_n=True).apply_simple(input=d)
+    assert len(out) == 5
+    
+    out = tablefilters.RandomSubset(n_to_select=150,
```
Comment (David-Baddeley) [inline tests/PYME/recipes/test_tablefilters.py]:
```suggestion
    out = tablefilters.RandomSubset(num_to_select=150,
```

**Pull request merged**